### PR TITLE
batch UI updates on a per frame basis

### LINKED ISSUE #6909: TypeError in the console
### Describe the bug

Only happens when using custom components. This is the component_loader.

[pngwn/gradio_imageslider#7](https://github.com/pngwn/gradio-imageslider/issues/7)

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

Use any custom component in build mode.

### Screenshot

-

### Logs

```shell
-
```

### System Info

```shell
-
```

### Severity

I can work around it

### LINKED ISSUE #7217: Fix rendering performance
### Describe the bug

Render performance is not good

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr

```

### Screenshot

Yes

### Logs

```shell
Yes
```

### System Info

```shell
All of them
```

### Severity

Blocking usage of gradio

### LINKED ISSUE #7633: accordion launches as open even when open = False since gradio 4.20
### Describe the bug

congrats for the work on gradio 4.20!!

a bug seems to have appeared: 
accordion launches as open even when open = False since gradio 4.20
parameter open = False seems not to work anymore

I didn't have the bug in gradio 4.17

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

 accordion = gr.Accordion(
            label="liste des fichiers",
            visible=True,
            open=False,
 )

### Screenshot

_No response_

### Logs

_No response_

### System Info

```shell
gradio 4.20

safari, chrome, iOS, mac : the bug is everywhere
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by pngwn**
## Description

This PR refactors component loading and state updates for three reasons:

- simplicity - isolating the component tree construction + tree updates makes the code easier to test and easier to reason about.
- correctness - this refactor includes some changes that ensure that we update the global state store when component props change from within the component. Previously, only initial props or props updated via a predict function where correctly updated.
- performance - by creating an abstracted update mechanism, we have more control over how and when we update state.

Performance is comparable to the current release but without the need to skip frames when rendering (the throttle has been removed), instead DOM updates are bound to a single frame.

Compare the following demos to illustrate:

This PR: https://pngwn-test-perf-pr.hf.space/
Current release: https://pngwn-test-perf-4-19.hf.space/

Locally both of these demos basically are pretty much in sync with the server updates, hosted versions have latency to account for.

For reference, this was the demo prior to @aliabid94's fix: https://pngwn-test-perf-4-15.hf.space/

**layout store**

This PR adds a 'layout store', this is a svelte store that is responsible for initialising the component tree (the logic that used to be in `Blocks.svelte`) as well as updating the tree (which contains all props for each component).

The tree construction has been refactored a little but is functionally the same, we iterate through the `layout` tree sent from the server, processing the data as necessary and loading all components.

The component loading specifically has been modified slightly, we kick off the component loading immediately, in parallel, and only await the imports later. I've also changed the component loader to cache requests and returned that cached request if the same promise is requested again, so we don't need to cache in the app itself. A bunch of unneeded code and structures have also been removed.

Updates have changed significantly, although functionally identical. Instead of processing every request as it comes in, we  do the following:

- if an update has not been scheduled then we schedule one.
- We push update transaction to an array (`{ id: number; prop: string; value: any }`)
- We do not update the global state

Scheduled updates will be run on the next frame, updating prior to this is pointless and can cause the UI to become unresponsive if updates are far enough apart not to get batched by svelte but close enough to hog the main thread. When the scheduled update is run we 'flush' the pending transactions. Transactions are run in the order they are received and synchronously, we don't try to do anything clever but dropping any, so state is guaranteed to be correct. As transactions are updated synchronously, we can also guarantee that no new transactions will be pushed to the pending transactions array while we are processing them. That will only happen after we have flushed all pending transactions.

This has the benefit that even if we update the same state multiple times, it will only result a single DOM update per frame, which is the most we ever really want to update things. Transactions are also immutable, the value being modified is always replaced with a new value and a new structure if appropriate and not mutated. This is important.

**immutable updates**

Component prop updates (both `value` and all other props)  must not be immutable. This is handled by the layout store, so other changes shouldn't be necessary. I have tested a variety of our own components as well as some custom component to ensure this doesn't cause any issues and i haven't noticed anything so far. But please test a few to see if you notice any issues, I will also continue to test.

What this means in simple terms is that updates from within a component may need to be adjusted or if everything is already fine, then we need to be aware of this in the future.

If we had some state like this:

```js
let my_state = {
  name: 'pngwn', 
  location: 'the beach'
}
```

Then we need to make sure that we don't just mutate the object but instead return a new one. 

Instead of doing this:

```js
my_state.location = 'the sea'
```

We need to do this:

```js
my_state = { ...my_state, location: 'the sea' }
```

This is a good practice anyway but is a requirement moving forward. We only care about the top level structure here, so spreading objects and arrays into a new one (which is essentially a shallow clone), is fine.

closes #7217. closes #6909. closes #7633

### CODE REVIEW

**Comment by pngwn** [comment]
works again now.

**Comment by pngwn** [comment]
And i've update the pr demo I linked above to use the latest gradio build for this PR.

**Comment by abidlabs** [comment]
Thanks @pngwn for this great PR! 

I just noticed an issue when working with `gr.DataFrame`:

1. If you run `demo/matrix_transpose/run.py`, you'll notice that the dataframe header and row alignment is off:

image

2. If you run `demo/kitchen_sink/run.py` and click on the first example, and then re-submit it, you'll see this error:

```
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/internals/construction.py", line 875, in to_arrays
    content, columns = _finalize_columns_and_data(arr, columns, dtype)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/internals/construction.py", line 972, in _finalize_columns_and_data
    raise ValueError(err) from err
```

To be honest, I don't understand how this is connected with the changes in this PR, but I've confirmed these errors are not present on `main`

**Comment by pngwn** [comment]
@abidlabs Should be fixed now. There were two issues:

- We weren't updating all of our internal header state, so it was getting reset to an old value. I've tweaked this to make sure we always have the correct header state internally and externally.
- Updates were sometimes not flushed before an API request was made. I've made sure API requests will wait for updates to settle before triggering an API request (ensuring the correct data is sent).

**Comment by abidlabs** [comment]
@pngwn I'm actually still seeing both issues when I test locally or on Spaces: https://huggingface.co/spaces/gradio-pr-deploys/pr-7564-all-demos

Perhaps some changes weren't pushed?

**Comment by pngwn** [comment]
My fix broke something else (parallel `.then` functions) just figuring it out.

**Comment by pngwn** [comment]
> @pngwn I'm actually still seeing both issues when I test locally or on Spaces: https://huggingface.co/spaces/gradio-pr-deploys/pr-7564-all-demos

> Perhaps some changes weren't pushed?

Weird, will check again locally. Maybe I messed something up.

**Comment by pngwn** [comment]
Will see what happens when the preview builds.

**Comment by pngwn** [comment]
Okay I understand now. I fixed it which broke the styled dataframes. fixing the styled dataframes brought the issue back.

Will find a better fix for the styled dataframes.

**Comment by pngwn** [comment]
I am caught in a loop where if i fix one bug another one comes back, and so on. 

I'll look at this again tomorrow with fresh eyes, untainted by safari.

**Comment by pngwn** [comment]
Okay, i think i fixed it now! 

We had an issue with the header prop not being updated correctly, and that overwriting the header key in the value.

**Review by abidlabs — APPROVED** [review]
LGTM @pngwn! Very nice PR. All the dataframe issues are fixed, and I tested a bunch of other demos and they are all working and quite snappily too.

File: js/dataframe/shared/Table.svelte [hunk 1]
```diff
@@ -41,21 +40,10 @@
 	export let column_widths: string[] = [];
 
 	let selected: false | [number, number] = false;
-	let display_value: string[][] | null = value?.metadata?.display_value ?? null;
-	let styling: string[][] | null = value?.metadata?.styling ?? null;
+	export let display_value: string[][] | null = null;
+	export let styling: string[][] | null = null;
```
[the remark this answers is not in the document]
Comment (pngwn) [reply js/dataframe/shared/Table.svelte:44]: I changed things around a little, we extract these values from `value` in `Index.svelte` now and pass them down.

**Comment by pngwn** [comment]
Thanks folks! Will take a look at that test, seems to be failing pretty consistently now.

**Pull request merged**